Homeworks/01-BWT/BurrowsWheelerTransformer/BurrowsWheelerTransformer/BWT.cs

```diff
@@ -0,0 +1,169 @@
+
+            for (var i = 0; i < rotations.Length; ++i)
+            {
+                var rotation = rotations[i];
+                if (rotation.EndPosition == 0)
+                {
+                    endPosition = i;
+                }
+
+                lastColumn[i] = rotation.Permutation[(((rotation.EndPosition - 1) % rotations.Length) + rotations.Length) % rotations.Length];
+            }
+
+            var resultingRotation = new Rotation(lastColumn.ToString(), endPosition);
+
+            return resultingRotation;
+        }
+
+        private static Rotation[] GenerateRotations(string input)
+        {
+            var rotations = new Rotation[input.Length];
+
+            for (var i = 0; i < input.Length; ++i)
+            {
+                rotations[i] = new Rotation(input, i);
+            }
+
+            return rotations;
+        }
+
+        private static Dictionary<char, int> GetDictionaryOfFirstOccurrences(Dictionary<char, int> frequencies)
+        {
+            var sortedSymbols = new List<char>(frequencies.Keys);
+            sortedSymbols.Sort();
+
+            var firstOccurrence = new Dictionary<char, int>();
+            var cumulativeCount = 0;
+
+            foreach (var symbol in sortedSymbols)
+            {
+                firstOccurrence[symbol] = cumulativeCount;
+                cumulativeCount += frequencies[symbol];
+            }
+
+            return firstOccurrence;
+        }
+
+        private static (Dictionary<char, int> Frequencies, int[] Rank) CountNumberOfEachSymbolAndItsRank(string input)
+        {
+            var rank = new int[input.Length];
+            var frequencies = new Dictionary<char, int>();
+            for (var i = 0; i < input.Length; i++)
+            {
+                var symbol = input[i];
+
+                if (!frequencies.TryAdd(symbol, 1))
+                {
+                    rank[i] = frequencies[symbol];
+                    frequencies[symbol]++;
+                }
+                else
+                {
+                    rank[i] = 0;
+                }
+            }
+
+            return (frequencies, rank);
+        }
+
+        private readonly struct Rotation(string stringForCyclicShifts, int endPosition) : IComparable<Rotation>
+        {
+            public string Permutation { get; } = stringForCyclicShifts;
+
+            public int EndPosition { get; } = endPosition;
+
+            public int CompareTo(Rotation other)
+            {
+                if (string.IsNullOrEmpty(this.Permutation))
+                {
+                    return string.IsNullOrEmpty(other.Permutation) ? 0 : -1;
+                }
+
+                if (string.IsNullOrEmpty(other.Permutation))
+                {
+                    return 1;
+                }
+
+                var i = this.EndPosition % this.Permutation.Length;
+                var j = other.EndPosition % other.Permutation.Length;
+
+                foreach (var unused in this.Permutation)
```

Comment:
В таких случаях принято использовать underscore:
```suggestion
                foreach (var _ in this.Permutation)
```
И это делается не только для красоты: компилятор [может это соптимизировать](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/tokens/discard).